File: app/src/main/java/br/com/jonathanzanella/myexpenses/source/ListSourceActivity.java
package br.com.jonathanzanella.myexpenses.source;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.TextView;

import br.com.jonathanzanella.myexpenses.R;
import br.com.jonathanzanella.myexpenses.views.BaseActivity;
import butterknife.BindView;

public class ListSourceActivity extends BaseActivity implements SourceAdapterCallback {
	public static final String KEY_SOURCE_SELECTED_UUID = "KeySourceSelectUuid";

	@BindView(R.id.act_sources_list)
	RecyclerView sources;
	@BindView(R.id.act_sources_list_empty)
	TextView emptyListView;

	@Override
	protected void onCreate(@Nullable Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_list_source);
		setTitle(R.string.select_source_title);
	}

	@Override
	protected void onPostCreate(Bundle savedInstanceState) {
		super.onPostCreate(savedInstanceState);

		SourceAdapter adapter = new SourceAdapter();
		adapter.setCallback(this);

		sources.setAdapter(adapter);
		sources.setHasFixedSize(true);
		sources.setLayoutManager(new GridLayoutManager(this, 2));
		sources.setItemAnimator(new DefaultItemAnimator());

		emptyListView.setVisibility(adapter.getItemCount() == 0 ? View.VISIBLE : View.GONE);
	}

	@Override
	public void onSourceSelected(Source source) {
		Intent i = new Intent();
		i.putExtra(KEY_SOURCE_SELECTED_UUID, source.getUuid());
		setResult(RESULT_OK, i);
		finish();
	}
}

File: app/src/main/java/br/com/jonathanzanella/myexpenses/source/ListSourceActivity.kt
package br.com.jonathanzanella.myexpenses.source

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.DefaultItemAnimator
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.View
import android.widget.TextView
import br.com.jonathanzanella.myexpenses.R
import br.com.jonathanzanella.myexpenses.views.anko.*
import org.jetbrains.anko.*

class ListSourceActivity : AppCompatActivity(), SourceAdapterCallback {
    private val ui = ListSourceActivityUi()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        ui.setContentView(this)

        ui.toolbar.title = getString(R.string.select_source_title)
        ui.toolbar.setup(this)
    }

    override fun onPostCreate(savedInstanceState: Bundle?) {
        super.onPostCreate(savedInstanceState)

        val adapter = SourceAdapter()
        adapter.setCallback(this)

        ui.sources.adapter = adapter
        ui.sources.setHasFixedSize(true)
        ui.sources.layoutManager = GridLayoutManager(this, 2)
        ui.sources.itemAnimator = DefaultItemAnimator()

        ui.emptyListView.visibility = if (adapter.itemCount == 0) View.VISIBLE else View.GONE
    }

    override fun onSourceSelected(source: Source) {
        val i = Intent()
        i.putExtra(KEY_SOURCE_SELECTED_UUID, source.uuid)
        setResult(Activity.RESULT_OK, i)
        finish()
    }

    companion object {
        val KEY_SOURCE_SELECTED_UUID = "KeySourceSelectUuid"
    }
}

class ListSourceActivityUi : AnkoComponent<ListSourceActivity> {
    lateinit var toolbar : TemplateToolbar
    lateinit var sources : RecyclerView
    lateinit var emptyListView : TextView

    override fun createView(ui: AnkoContext<ListSourceActivity>) = with(ui) {
        verticalLayout {
            toolbar = toolbarTemplate {}
            sources = recyclerView {}
            emptyListView = emptyListMessageView { text = resources.getString(R.string.message_no_sources) }
        }.applyRecursively(::applyTemplateViewStyles)
    }
}